File: golang/internal/steps/step_configure.go
package steps

/* ---------------------------------------------------------------- *
 * IMPORTS
 * ---------------------------------------------------------------- */

import (
	"fmt"
	"io/ioutil"
	"path/filepath"
	"reflect"
	"strings"

	"gopkg.in/yaml.v3"

	"phpytex/internal/core/logging"
	"phpytex/internal/core/utils"
	"phpytex/internal/setup/appconfig"
	"phpytex/internal/setup/userconfig"
	"phpytex/pkg/re"
)

/* ---------------------------------------------------------------- *
 * METHOD configure
 * ---------------------------------------------------------------- */

func Configure(fnameConfig string) error {
	var err error
	var config = &(userconfig.UserConfig{})

	logging.LogInfo("READ CONFIG STARTED")

	err = getUserConfig(fnameConfig, config)
	if err != nil {
		return fmt.Errorf(fmt.Sprintf("Could not read config file \033[1m%s\033[0mor its contents were invalid.", fnameConfig))
	}

	err = setHeaderConfig(config.Header)
	if err != nil {
		return err
	}

	err = setCompileConfig(config.Compile)
	if err != nil {
		return err
	}

	err = setStampConfig(config.Stamp)
	if err != nil {
		return err
	}

	err = setParamsConfig(config.Parameters)
	if err != nil {
		return err
	}

	setConfigFilesAndFolders(config.Tree)

	logging.LogInfo("READ CONFIG COMPLETE")
	return nil
}

/* ---------------------------------------------------------------- *
 * PRIVATE METHODS
 * ---------------------------------------------------------------- */

func getUserConfig(fnameConfig string, config *userconfig.UserConfig) error {
	var err error
	var contents []byte
	if fnameConfig == "" {
		fnameConfig, err = utils.GetFirstFileByPattern(appconfig.Parameters.PathRoot.GetValue(), appconfig.Parameters.PatternConfig.GetValue())
	}
	if err != nil {
		return err
	}
	contents, err = ioutil.ReadFile(fnameConfig)
	if err != nil {
		return err
	}
	err = yaml.Unmarshal(contents, config)
	if err != nil {
		return err
	}
	userconfig.HandleMissingSections(config)
	userconfig.HandleBackwardsCompatibility(config)
	userconfig.HandleMissingKeys(config)
	return nil
}

func setHeaderConfig(config *userconfig.HeaderConfig) error {
	appconfig.Parameters.OptionIgnore.SetValueFromPtr(config.Ignore)
	return nil
}

func setCompileConfig(config *userconfig.CompileConfig) error {
	var err error = nil
	var (
		comments    interface{}
		valueString string
		valueBool1  bool
		valueBool2  bool
	)
	var (
		indentSymb string
		spaces     int
	)
	var (
		root       string
		fileStart  string
		fileOutput string
		file       string
	)

	appconfig.Parameters.OptionLegacy.SetValueFromPtr(config.Legacy)
	appconfig.Parameters.OptionDebug.SetValueFromPtr(config.Debug)
	appconfig.Parameters.OptionCompileLatex.SetValueFromPtr(config.Compile)
	appconfig.Parameters.OptionInsertBib.SetValueFromPtr(config.InsertBib)
	appconfig.Parameters.OptionShowTree.SetValueFromPtr(config.ShowTree)

	comments = *config.Comments
	if reflect.TypeOf(comments).Kind() == reflect.String {
		valueString = comments.(string)
		valueBool1 = re.Matches(`(?i)(^(auto|default)$)`, valueString)
		valueBool2 = re.Matches(`(?i)(^(on|default)$)`, valueString) || !re.Matches(`(?i)(^(|off)$)`, valueString)
	} else if reflect.TypeOf(comments).Kind() == reflect.Bool {
		// NOTE: if boolean value used, then user wants all on or all off, thus [auto] = false
		valueBool1 = false
		valueBool2 = comments.(bool)
	} else {
		valueBool1 = true
		valueBool2 = true
	}
	appconfig.Parameters.OptionCommentsAuto.SetValue(valueBool1)
	appconfig.Parameters.OptionCommentsOn.SetValue(valueBool2)

	appconfig.Parameters.MaxLength.SetValueFromPtr(config.MaxLength)
	appconfig.Parameters.Seed.SetValueFromPtr(config.Seed)

	if *config.Tabs {
		appconfig.Parameters.IndentCharacter.SetValue("\t")
		appconfig.Parameters.IndentCharacterRe.SetValue(`\t`)
	} else {
		spaces = *config.Spaces
		indentSymb = strings.Repeat(" ", spaces)
		appconfig.Parameters.IndentCharacter.SetValue(indentSymb)
		appconfig.Parameters.IndentCharacterRe.SetValue(indentSymb)
	}
	indentSymb = appconfig.Parameters.IndentCharacter.GetValue()
	if utils.LengthOfWhiteSpace(indentSymb) == 0 {
		return fmt.Errorf("Indentation symbol cannot be the empty string!")
	}

	appconfig.Parameters.Offset.SetValueFromPtr(config.Offset)
	if *config.Legacy && config.Offset == nil {
		appconfig.Parameters.Offset.SetValue(indentSymb)
	}

	root = appconfig.Parameters.PathRoot.GetValue()
	fileStart, err = utils.FormatPath(*config.Root, root, false)
	if err != nil {
		return fmt.Errorf(fmt.Sprintf("Could not process input path \033[1m%[1]s\033[0m.", *config.Root))
	}
	fileOutput, err = utils.FormatPath(*config.Output, root, false, nil, utils.StringToPtr(".tex"))
	if err != nil {
		return fmt.Errorf(fmt.Sprintf("Could not process output path \033[1m%[1]s\033[0m.", *config.Root))
	}
	if filepath.Dir(fileOutput) != root {
		return fmt.Errorf("The output file can only be set to be in the root directory!'")
	}
	if fileStart == fileOutput {
		return fmt.Errorf("The output and start ('root'-attribute in config) paths must be different!")
	}

	appconfig.Parameters.FileStart.SetValue(fileStart)
	appconfig.Parameters.FileOutput.SetValue(fileOutput)

	file, err = utils.FormatPath("phpytex_transpiled.py", root, false)
	if err != nil {
		return err
	}
	appconfig.Parameters.FileTranspiled.SetValue(file)
	return nil
}

func setStampConfig(config *userconfig.StampFileConfig) error {
	var err error = nil
	var (
		root string
		file string
	)

	appconfig.Parameters.WithFileStamp.SetValue(false)
	if config != nil && config.Options != nil && len(*config.Options) != 0 {
		root = appconfig.Parameters.PathRoot.GetValue()
		file = *config.File
		file, err = utils.FormatPath(file, root, false)
		if err != nil {
			return err
		}
		appconfig.Parameters.WithFileStamp.SetValue(true)
		appconfig.Parameters.FileStamp.SetValue(file)
		appconfig.Parameters.OptionOverwriteStamp.SetValueFromPtr(config.Overwrite)
		appconfig.DictionaryStamp.SetValues(config.Options)
	}
	return err
}

func setParamsConfig(config *userconfig.ParametersFileConfig) error {
	var err error = nil
	var (
		root string
	)
	var (
		modulename string
		path       string
	)

	appconfig.Parameters.WithFileParamsPy.SetValue(false)
	appconfig.Parameters.OptionOverwriteParams.SetValueFromPtr(config.Overwrite)
	appconfig.DictionaryParams.SetValues(config.Options)

	root = appconfig.Parameters.PathRoot.GetValue()
	if config != nil && config.Options != nil && len(*config.Options) != 0 {
		modulename = utils.PtrToString(config.File, "")
		if re.Matches(`^[^\.\s]*(\.[^\.\s]*)+$`, modulename) {
			path = re.Sub(`([^\.]+)\.`, `$1/`, modulename)
			path = path + ".py"
		} else {
			return fmt.Errorf("\033[1mparameters > file\033[0m option must by a python-like import path (relative to the root of the project).")
		}
		appconfig.Parameters.WithFileParamsPy.SetValue(true)
		appconfig.Parameters.ImportParamsPy.SetValue(modulename)
		path, err = utils.FormatPath(path, root, false)
		if err != nil {
			return err
		}
		appconfig.Parameters.FileParamsPy.SetValue(path)
	}

	return err
}

func setConfigFilesAndFolders(config *userconfig.TreeConfig) {
	appconfig.ProjectTree = config
}

File: src/steps/step_configure.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# IMPORTS
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

from src.local.misc import *;
from src.local.system import *;
from src.local.typing import *;

from src.core.log import *;
from src.core.utils import createNewFileName;
from src.core.utils import formatPath;
from src.core.utils import getAttribute;
from src.core.utils import getFilesByPattern;
from src.core.utils import lengthOfWhiteSpace;
from src.core.utils import readYamlFile;
from src.core.utils import restrictDictionary;
from src.core.utils import toPythonKeysDict;
from src.customtypes.exports import ProjectTree;
from src.setup import appconfig;
from src.setup.userconfig import setupYamlReader;

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# GLOBAL VARIABLES
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# METHOD: step get config
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

def step(fnameConfig: str, extra_parameters: dict):
    logInfo('READ CONFIG STARTED');
    ## get configuration file
    config = getPhpytexConfig(fnameConfig);
    ## get main parts of config
    config_compile = getAttribute(config, 'compile', 'options', expectedtype=dict, default=None) \
                     or getAttribute(config, 'compile', expectedtype=dict, default={});
    config_compile = preProcessCompileConfig({ **restrictDictionary(config, ['ignore']), **config_compile });
    config_stamp = getAttribute(config, 'stamp', expectedtype=dict, default={});
    config_parameters: dict = getAttribute(config, 'parameters', expectedtype=dict, default={});
    config_parameters['options'] = {**config_parameters.get('options', {}), **extra_parameters};

    ## set app config
    setCompileConfig(**config_compile);
    setStampConfig(**toPythonKeysDict(config_stamp));
    setParamsConfig(**toPythonKeysDict(config_parameters));
    setConfigFilesAndFolders(toPythonKeysDict(config));

    logInfo('READ CONFIG COMPLETE');
    return;

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# SECONDARY METHODS
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

def getPhpytexConfig(fnameConfig: str) -> Dict[str, Any]:
    setupYamlReader();
    try:
        if not isinstance(fnameConfig, str) or fnameConfig == '':
            fnameConfig = getFilesByPattern(
                path        = appconfig.getPathRoot(),
                filepattern = appconfig.getPatternConfig()
            )[0];
    except:
        raise Exception('Could not find or read any phpytex configuration files.');
    return readYamlFile(fnameConfig);

def preProcessCompileConfig(config: Dict[str, Any]) -> Dict[str, Any]:
    return dict(
        ignore     = getAttribute(config, 'ignore', expectedtype=bool, default=False),
        legacy     = getAttribute(config, 'legacy', expectedtype=bool, default=False),
        startfile  = getAttribute(config, ['root', 'input'], expectedtype=str),
        outputfile = getAttribute(config, 'output', expectedtype=str, default='main.tex'),
        debug      = getAttribute(config, 'debug', expectedtype=bool, default=False),
        compile    = getAttribute(config, ['compile-latex', 'compile'], expectedtype=bool, default=False),
        insert_bib = getAttribute(config, 'insert-bib', expectedtype=bool, default=True),
        backend_bib = getAttribute(config, 'backend-bib', expectedtype=str, default='bibtex'),
        comments   = getAttribute(config, 'comments', expectedtype=(str,bool), default='auto'),
        show_tree  = getAttribute(config, ['show-structure', 'show-tree'], expectedtype=bool, default=False),
        max_length = getAttribute(config, 'max-length', expectedtype=int, default=10000),
        tabs       = getAttribute(config, 'tabs', expectedtype=bool, default=False),
        spaces     = getAttribute(config, 'spaces', expectedtype=int, default=4),
        ## NOTE: Do not force seed to be set if not given. Allow user to decide to NOT seed the rng.
        seed       = getAttribute(config, 'seed', expectedtype=int, default=None),
        offset     = getAttribute(config, 'offset', expectedtype=str, default=''),
        align      = getAttribute(config, 'align', expectedtype=bool, default=True),
    );

def setCompileConfig(
    ignore:     bool,
    legacy:     bool,
    startfile:  str,
    outputfile: str,
    debug:      bool,
    compile:    bool,
    insert_bib: bool,
    backend_bib: str,
    comments:   Union[str, bool],
    show_tree:  bool,
    max_length: int,
    tabs:       bool,
    spaces:     int,
    seed:       Any,
    offset:     str,
    align:      bool,
):
    root = appconfig.getPathRoot();
    appconfig.setOptionLegacy(legacy);
    appconfig.setOptionIgnore(ignore);
    appconfig.setOptionDebug(debug);
    appconfig.setOptionCompileLatex(compile);
    appconfig.setOptionInsertBib(insert_bib);
    appconfig.setOptionBackendBib(backend_bib);
    appconfig.setOptionShowTree(show_tree);
    if isinstance(comments, str):
        appconfig.setOptionCommentsAuto(comments in [ 'auto', 'default' ]);
        appconfig.setOptionCommentsOn(comments in [ 'on', 'default' ] or not comments in [ 'off' ]);
    else:# elif isinstance(comments, bool):
        appconfig.setOptionCommentsAuto(False);
        appconfig.setOptionCommentsOn(comments);
    if isinstance(seed, int):
        appconfig.setSeed(seed);
    appconfig.setOffsetSymbol(offset);
    appconfig.setOptionAlign(align);

    appconfig.setMaxLengthOutput(max_length);
    if tabs:
        appconfig.setIndentCharacter('\t');
        appconfig.setIndentCharacterRe(r'\t');
    else:
        appconfig.setIndentCharacter(' '*spaces);
        appconfig.setIndentCharacterRe(' '*spaces);
    indentsymb = appconfig.getIndentCharacter();
    if lengthOfWhiteSpace(indentsymb) == 0:
        raise AttributeError('Indentation symbol cannot be the empty string!');
    if legacy:
        appconfig.setOffsetSymbol(indentsymb);

    fileStart = formatPath(startfile, root=root, relative=False);
    fileOutput = formatPath(outputfile, root=root, relative=False, ext_if_empty='.tex');

    assert os.path.dirname(fileOutput) == root, 'The output file can only be set to be in the root directory!';
    assert not (fileStart == fileOutput), 'The output and start (\'root\'-attribute in config) paths must be different!';

    appconfig.setFileStart(fileStart);
    appconfig.setFileOutput(fileOutput);

    # file = createNewFileName(dir=root, nameinit='phpytex_transpiled.py', namescheme='phpytex_transpiled_{}.py');
    file = 'phpytex_transpiled.py';
    appconfig.setFileTranspiled(formatPath(file, root=root, relative=False));
    return;

def setStampConfig(
    file: str = '',
    overwrite: bool = True,
    options: Dict[str, Any] = dict()
):
    root = appconfig.getPathRoot();
    if not isinstance(file, str) or file == '':
        file = 'stamp.tex';
        file = os.path.relpath(path=file, start=root);

    if not isinstance(options, dict) or len(options) == 0:
        appconfig.setWithFileStamp(False);
    else:
        appconfig.setWithFileStamp(True);
        appconfig.setFileStamp(formatPath(file, root=root, relative=False));
        appconfig.setOptionOverwriteStamp(overwrite);
        appconfig.setDictionaryStamp(options);
    return;

def setParamsConfig(
    file: str = '',
    overwrite: bool = True,
    options: Dict[str, Any] = dict(),
):
    appconfig.setOptionOverwriteParams(overwrite);
    appconfig.setDictionaryParams(options);

    root = appconfig.getPathRoot();
    modulename = file if isinstance(file, str) else '';
    if re.match(r'^[^\.\s]*(\.[^\.\s]*)+$', file):
        path = re.sub(r'([^\.]+)\.', r'\1/', file) + '.py';
    else:
        logWarn('\033[1mparameters > file\033[0m option must by a python-like import path (relative to the root of the project).');
        path = 'parameters.py';
        path = os.path.relpath(path, root);
    modulename = re.sub(r'\/', '.', os.path.splitext(path)[0]);

    if not isinstance(options, dict) or len(options) == 0:
        appconfig.setWithFileParamsPy(False);
    else:
        appconfig.setWithFileParamsPy(True);
        appconfig.setImportParamsPy(modulename);
        appconfig.setFileParamsPy(formatPath(path, root=root, relative=False));
    return;

def setConfigFilesAndFolders(config: Dict[str, Any]):
    appconfig.setProjectTree(ProjectTree(**config));
    return;
